Minimum Working Dockerfile

NoRePercussions opened the pull request:
A mostly-functional Docker environment for tracker.

Some things this doesn't include:

- PDF generation (grover/puppeteer/chromium)
- Production environment
- Static asset serving documentation

DaAwesomeP commented on Dockerfile:
```diff
@@ -31,7 +35,7 @@ RUN curl -sL https://github.com/nodenv/node-build/archive/master.tar.gz | tar xz
 
 # Install application gems
-COPY Gemfile Gemfile.lock ./
+COPY Gemfile Gemfile.lock .ruby-version ./
```
Is this necessary since the container already has the version baked in?

NoRePercussions replied:
Bundler is upset if I don't include it. I agree that it isn't great to have it duplicated, but as we are also supporting local development for the time being, I'm not a huge fan of any method to sync it either.

DaAwesomeP replied:
Ah, what if you remove this line:
https://github.com/ABTech/tracker/blob/d9cb1c243f0540ce07857f3197bb24a29054d624/Gemfile#L3

I haven't seen this line in other Rails projects. It's probably a remnant of the past.

DaAwesomeP replied:
Now that I think of it--I have run into this before and we should definitely remove this line. This line _enforces_ a specific Ruby version which is really annoying if you are developing on a machine with a different version (and without `rvm`).

DaAwesomeP commented on Dockerfile:
```diff
@@ -57,12 +61,8 @@ RUN bundle exec bootsnap precompile app/ lib/
-      mv config/credentials/sample.key config/master.key && \
-      SECRET_KEY_BASE_DUMMY=1 ./bin/rails assets:precompile && \
-      mv config/credentials.yml.enc.backup config/credentials.yml.enc && \
+      SECRET_KEY_BASE_DUMMY=1 RAILS_ENV=development ./bin/rails assets:precompile && \
```
We should add a `TODO` comment to revisit this workaround and also make sure we don't leave the hard-coded `development` env in there.

DaAwesomeP commented on Dockerfile:
```diff
@@ -19,10 +19,14 @@ FROM base as build
+    build-essential \
+    curl \
+    libmariadb-dev  `# Needed by ActiveRecord` \
+    chromium  # Needed by grover
```
So if I remember correctly, Grover/Puppeteer fetches its own binary for Chromium and we just need to make sure it has all of the dependencies it needs (but not Chromium itself, which depends on X window packages).

NoRePercussions replied:
It doesn't on aarch64. I'm also concerned about versioning issues with this. It might be safer, though less convenient, to just restrict it to x86-64.

DaAwesomeP replied:
Yeah if you use the Chromium provided by Puppeteer then you are guaranteed that the version will match. Also _most_ of the functionality will work fine on aarch64, just not PDF generation. I think it is fine to restrict it to x86. I also definitely would not want all of the X dependencies that come with the Chromium package in the Docker image.

NoRePercussions replied:
Puppeteer doesn't install at all without chromium. That can be switched off with an environment flag, but then it wouldn't bring its own binary. I also don't like the idea of conditional builds nor keeping differing environments on different platforms, so I'd also like to just restrict platform.

DaAwesomeP commented on Dockerfile:
```diff
@@ -19,10 +19,14 @@ FROM base as build
 # Install packages needed to build gems and node modules
 RUN apt-get update -qq && \
-    apt-get install --no-install-recommends -y build-essential curl git libpq-dev libvips node-gyp pkg-config python-is-python3
+    apt-get install --no-install-recommends -y \
+    build-essential \
+    curl \
+    libmariadb-dev  `# Needed by ActiveRecord` \
```
I know `libvips` is used by ActiveRecord. `node-gyp` and `pkg-config` are used for building native Node modules (possibly the python one too), which are a part of the asset pipeline IIRC.

I'm confused how Rails recommends the SQL client lib for MariaDB, though for our version we definitely needs `libmariadb-dev`. See the last link below, but they don't install a native package for MariaDB (or maybe install the MySQL client which works too?). I suppose we could run `rails new` and tell it use to MariaDB and see what it spits out.

Ideally we stick to the Rails-suggested Dockerfile templates as much as possible, This is to make Rails version upgrades much, much easier and so won't have unexpected issues (i.e. implicit dependency provided by another package changes). When we run `rails app:update` after hitting Rails 7 it will suggest changes to this file.

See these files:
https://github.com/rails/rails/blob/3769f1eee6a1ec4f6438155d67477db6d192577f/railties/lib/rails/generators/app_base.rb#L589
https://github.com/rails/rails/blob/3769f1eee6a1ec4f6438155d67477db6d192577f/railties/lib/rails/generators/rails/app/templates/Dockerfile.tt#L18
https://github.com/rails/rails/blob/3769f1eee6a1ec4f6438155d67477db6d192577f/railties/lib/rails/generators/database.rb#L173

NoRePercussions replied:
I didn't have any issues during build or basic testing after removing these, but I haven't run further tests. I'll restore it to Rails defaults for compatibility.

I'm not sure if Rails needs it for connecting to MariaDB as `mariadb://`. However our development steps use MySQL, and we use the mysql2 gem, which relies on mariadb dev tools.

DaAwesomeP commented on config/puma.rb:
```diff
@@ -15,7 +15,7 @@
 # Specifies the `port` that Puma will listen on to receive requests; default is 3000.
 #
-port ENV.fetch("PORT") { 3000 }
+bind "tcp://0.0.0.0:80"
```
I think configurable port is a good thing, especially for development. Also for those not yet using Docker to develop, listening on 80 won't work unless running as root. See Rails default here (`rails app:update` will flag this one too):

https://github.com/rails/rails/blob/3769f1eee6a1ec4f6438155d67477db6d192577f/railties/lib/rails/generators/rails/app/templates/config/puma.rb.tt#L48

NoRePercussions replied:
I agree. The change is partly to bind to all interfaces, but I will fix it to bind to 3000 by default, configurable.

DaAwesomeP replied:
I believe you can still override Puma through other env variables run or just running the `puma` command directly. It's also probably not good to bind to `0.0.0.0` by default for development  (security). You also probably don't need to worry about binding in the Docker container.

NoRePercussions replied:
Ok, I see that there's an arg to do this. That seems better. Unfortunately in the container it binds to loopback and won't connect unless configured.

DaAwesomeP replied:
Try with `RAILS_ENV=production`. From the `rails server --help`:
```
Usage:
  rails server -u [thin/puma/webrick] [options]

Options:
  -e, [--environment=ENVIRONMENT]              # Specifies the environment to run this server under (test/development/production).
  -p, [--port=port]                            # Runs Rails on the specified port - defaults to 3000.
  -b, [--binding=IP]                           # Binds Rails to the specified IP - defaults to 'localhost' in development and '0.0.0.0' in other environments'.
...
```

DaAwesomeP replied:
Ah, makes sense

DaAwesomeP reviewed (changes requested):
This is great! Just a few notes.

I really prefer to stick to the Rails templates for the configs unless there is a strong reason to waiver. After updating Tracker through the entire chain from 5.0 to 6.1 (took at least a month), I don't wish on anyone to have to go through so so many changes that don't actually affect runtime and just things hard to understand (i.e. "was this a workaround" or "someone just made it different").

NoRePercussions commented on Dockerfile:
```diff
@@ -88,12 +93,13 @@ USER rails:rails
 RUN mkdir -p /srv/public \
- && chown -R rails:rails /srv/public
+ && chown -R rails:rails /srv/public \
+ && chmod +x -R /rails/bin
```
Yep, `docker-entrypoint` isn't in the right mode. Will fix and remove.

DaAwesomeP commented on Dockerfile:
```diff
@@ -95,5 +101,5 @@ USER rails:rails
 # Start the server by default, this can be overwritten at runtime
-EXPOSE 3000
-CMD ["./bin/rails", "server"]
+EXPOSE 80
```
```suggestion
EXPOSE 3000
```

NoRePercussions commented:
The lifecycle of the development containers seems pretty automatable. What do you think about including some `make` targets for development?

DaAwesomeP commented:
> The lifecycle of the development containers seems pretty automatable. What do you think about including some `make` targets for development?

We could, but ideally in the end I think it would condense down to:
```bash
docker compose up --file compose_dev.yaml
docker exec -it tracker rails c  # for initial member account
```
That way you don't even have to setup SQL. The `compose_dev.yaml` could build the repo locally and the `compose.yaml` (for production) would pull from the GitHub container registry.

> I'd also personally like to include a healthcheck, but I don't want to deviate too far from the rails template.

When we reach Rails 7 there is this: https://edgeapi.rubyonrails.org/classes/Rails/HealthController.html

NoRePercussions commented:
Great. I think I've addressed all comments, so what remains is whatever remaining comments you have.

It seems like after this PR, next steps in dockerizing may be
- Github actions for build check, run check
- Set up GHCR
- Setting up docker-compose dev env and everything that entails
- as well as continuing to work on salt interface for compose

DaAwesomeP commented:
Great! I have a standard GitHub action for dealing with GHCR I can push later tonight.

I think we should try to tackle #544 as a part of this. The current email process and Sphinx process are trivial to run in Docker Compose with the same container, but the Slack notification runs on a Systemd timer. That Systemd service _can_ just become a `docker exec` but it would be better to be self-contained.
